{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DSS-AZUL-INDEXING\n",
    "\n",
    "### This jupyter notebook is to do a couple testing indexing operations using the DataExtractor class and the FileIndexer class."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below here we import our modules and set up:\n",
    "* ElasticSearch Client\n",
    "* Dummy payload of event\n",
    "* Parse the bundle_uuid and the bundle_version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from elasticsearch import Elasticsearch\n",
    "from indexer.utils import DataExtractor\n",
    "from indexer.indexer import FileIndexer\n",
    "import json\n",
    "from pprint import pprint\n",
    "\n",
    "# Create an ElasticSearch client\n",
    "es = Elasticsearch([{'host': 'localhost', 'port': 9200}])\n",
    "# Sample event payload\n",
    "payload = { \"query\": { \"query\": { \"match_all\":{}} }, \"subscription_id\": \"ba50df7b-5a97-4e87-b9ce-c0935a817f0b\", \"transaction_id\": \"ff6b7fa3-dc79-4a79-a313-296801de76b9\", \"match\": { \"bundle_version\": \"2017-10-02T183005.116143Z\", \"bundle_uuid\": \"cd55a68f-fc98-4c6d-af5e-7df491775df4\" } }\n",
    "bundle_uuid = payload['match']['bundle_uuid']\n",
    "bundle_version = payload['match']['bundle_version']\n",
    "# Create "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will create an instance of the DataExtractor and use it to get the contents from the bundle referenced by the variable `payload`. We will be pulling from the AWS replica."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "#####################################################\n",
      "#                    PRINTING METADATA               #\n",
      "#####################################################\n",
      "{   'assay.json': {   'core': {   'accession': None,\n",
      "                                  'events': [   {   'endState': 'Validating',\n",
      "                                                    'originalState': 'Draft',\n",
      "                                                    'submissionDate': {   'date': '2017-10-02T18:13:29.336+0000'}},\n",
      "                                                {   'endState': 'Valid',\n",
      "                                                    'originalState': 'Validating',\n",
      "                                                    'submissionDate': {   'date': '2017-10-02T18:13:39.375+0000'}}],\n",
      "                                  'submissionDate': {   'date': '2017-10-02T18:10:53.603Z'},\n",
      "                                  'updateDate': {   'date': '2017-10-02T18:13:39.375Z'},\n",
      "                                  'uuid': '90dab527-46d0-4e81-8024-1b017189d8ab',\n",
      "                                  'validationState': 'Valid'},\n",
      "                      'id': 'ERR1631082',\n",
      "                      'rna': {   'library_construction': 'smart-seq2',\n",
      "                                 'primer': 'random',\n",
      "                                 'spike_in': 'ERCC',\n",
      "                                 'spike_in_dilution': 40000,\n",
      "                                 'strand': 'both'},\n",
      "                      'sample_id': 'HP1504901_J8',\n",
      "                      'seq': {   'ena_experiment': 'ERX1701415',\n",
      "                                 'ena_run': 'ERR1631082',\n",
      "                                 'instrument_model': 'Illumina HiSeq 2000',\n",
      "                                 'lanes': [   {   'number': 1,\n",
      "                                                  'r1': 'ERR1631082.fastq.gz'}],\n",
      "                                 'molecule': 'RNA',\n",
      "                                 'paired_ends': 'no'},\n",
      "                      'single_cell': {'cell_handling': 'FACS'}},\n",
      "    'project.json': {   'array_express_investigation': 'E-MTAB-5061',\n",
      "                        'core': {   'accession': None,\n",
      "                                    'events': [   {   'endState': 'Validating',\n",
      "                                                      'originalState': 'Draft',\n",
      "                                                      'submissionDate': {   'date': '2017-10-02T18:11:16.784+0000'}},\n",
      "                                                  {   'endState': 'Valid',\n",
      "                                                      'originalState': 'Validating',\n",
      "                                                      'submissionDate': {   'date': '2017-10-02T18:11:26.811+0000'}}],\n",
      "                                    'submissionDate': {   'date': '2017-10-02T18:10:36.660Z'},\n",
      "                                    'updateDate': {   'date': '2017-10-02T18:11:26.811Z'},\n",
      "                                    'uuid': '9a8c8be0-d806-4043-830b-49edf623263e',\n",
      "                                    'validationState': 'Valid'},\n",
      "                        'ddjb_trace': 'ERP017126',\n",
      "                        'description': 'We used single-cell RNA-sequencing to '\n",
      "                                       'generate transcriptional profiles of '\n",
      "                                       'endocrine and exocrine cell types of '\n",
      "                                       'the human pancreas. Pancreatic tissue '\n",
      "                                       'and islets were obtained from six '\n",
      "                                       'healthy and four T2D cadaveric donors. '\n",
      "                                       'Islets were cultured and dissociated '\n",
      "                                       'into single-cell suspension. Viable '\n",
      "                                       'individual cells were distributed via '\n",
      "                                       'fluorescence-activated cell sorted '\n",
      "                                       '(FACS) into 384-well plates containing '\n",
      "                                       'lysis buffer. Single-cell cDNA '\n",
      "                                       'libraries were generated using the '\n",
      "                                       'Smart-seq2 protocol. Gene expression '\n",
      "                                       'was quantified as reads per kilobase '\n",
      "                                       'transcript and per million mapped '\n",
      "                                       'reads (RPKM) using rpkmforgenes. '\n",
      "                                       'Bioinformatics analysis was used to '\n",
      "                                       'classify cells into cell types without '\n",
      "                                       'knowledge of cell types or prior '\n",
      "                                       'purification of cell populations. We '\n",
      "                                       'revealed subpopulations in endocrine '\n",
      "                                       'and exocrine cell types, identified '\n",
      "                                       'genes with interesting correlations to '\n",
      "                                       'body mass index (BMI) in specific cell '\n",
      "                                       'types and found transcriptional '\n",
      "                                       'alterations in T2D.  Complementary '\n",
      "                                       'whole-islet RNA-seq data have also '\n",
      "                                       'been deposited at ArrayExpress under '\n",
      "                                       'accession number E-MTAB-5060 '\n",
      "                                       '(http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-5060).',\n",
      "                        'experimental_design': [   'cell type comparison '\n",
      "                                                   'design',\n",
      "                                                   'disease state design'],\n",
      "                        'experimental_factor_name': ['cell type', 'disease'],\n",
      "                        'id': 'E-MTAB-5061',\n",
      "                        'mage_tab_version': 1.1,\n",
      "                        'related_projects': ['E-MTAB-5060'],\n",
      "                        'release_status': '2016-09-22',\n",
      "                        'submitter': {   'address': 'Nobels väg 3, 171 77, '\n",
      "                                                    'Stockholm, Sweden',\n",
      "                                         'city': 'Stockholm',\n",
      "                                         'country': 'Sweden',\n",
      "                                         'email': 'Athanasia.Palasantza@ki.se',\n",
      "                                         'institution': 'Department of Cell '\n",
      "                                                        'and Molecular Biology '\n",
      "                                                        '(CMB), Karolinska '\n",
      "                                                        'Institutet, '\n",
      "                                                        'Stockholm, Sweden.',\n",
      "                                         'name': 'Athanasia,,Palasantza',\n",
      "                                         'phone': '0046 8 5248 3986'},\n",
      "                        'supplementary_files': ['E-MTAB-5061.README.txt'],\n",
      "                        'title': 'Single-cell RNA-seq analysis of human '\n",
      "                                 'pancreas from healthy individuals and type 2 '\n",
      "                                 'diabetes patients'},\n",
      "    'sample.json': {   'biosd_sample': 'SAMEA4438090',\n",
      "                       'body_part': {   'name': 'islet of Langerhans',\n",
      "                                        'ontology': 'UBERON_0000006'},\n",
      "                       'core': {   'accession': None,\n",
      "                                   'events': [   {   'endState': 'Validating',\n",
      "                                                     'originalState': 'Draft',\n",
      "                                                     'submissionDate': {   'date': '2017-10-02T18:11:59.093+0000'}},\n",
      "                                                 {   'endState': 'Valid',\n",
      "                                                     'originalState': 'Validating',\n",
      "                                                     'submissionDate': {   'date': '2017-10-02T18:12:09.138+0000'}}],\n",
      "                                   'submissionDate': {   'date': '2017-10-02T18:10:44.554Z'},\n",
      "                                   'updateDate': {   'date': '2017-10-02T18:12:09.138Z'},\n",
      "                                   'uuid': '09e91896-e730-4a3c-baba-44884eb9506b',\n",
      "                                   'validationState': 'Valid'},\n",
      "                       'donor': {   'age': 23,\n",
      "                                    'age_unit': 'year',\n",
      "                                    'body_mass_index': 21.5,\n",
      "                                    'core': {   'accession': None,\n",
      "                                                'events': [   {   'endState': 'Validating',\n",
      "                                                                  'originalState': 'Draft',\n",
      "                                                                  'submissionDate': {   'date': '2017-10-02T18:11:16.797+0000'}},\n",
      "                                                              {   'endState': 'Valid',\n",
      "                                                                  'originalState': 'Validating',\n",
      "                                                                  'submissionDate': {   'date': '2017-10-02T18:11:28.861+0000'}}],\n",
      "                                                'submissionDate': {   'date': '2017-10-02T18:10:36.724Z'},\n",
      "                                                'updateDate': {   'date': '2017-10-02T18:11:28.861Z'},\n",
      "                                                'uuid': 'd39785ff-3683-4a74-ad45-3a7a36913a25',\n",
      "                                                'validationState': 'Valid'},\n",
      "                                    'disease': {'text': 'normal'},\n",
      "                                    'id': 'HP1504901',\n",
      "                                    'life_stage': 'adult',\n",
      "                                    'ncbi_taxon': 9606,\n",
      "                                    'sex': 'male',\n",
      "                                    'species': {'text': 'Homo sapiens'}},\n",
      "                       'donor_id': 'HP1504901',\n",
      "                       'ena_sample': 'ERS1349539',\n",
      "                       'id': 'HP1504901_J8',\n",
      "                       'name': 'pancreas islet normal',\n",
      "                       'organ': {   'ontology': 'UBERON_0001264',\n",
      "                                    'text': 'pancreas'},\n",
      "                       'protocol_ids': ['p1', 'p2', 'p3', 'p4'],\n",
      "                       'protocols': [   {   'core': {   'accession': None,\n",
      "                                                        'events': [   {   'endState': 'Validating',\n",
      "                                                                          'originalState': 'Draft',\n",
      "                                                                          'submissionDate': {   'date': '2017-10-02T18:11:16.774+0000'}},\n",
      "                                                                      {   'endState': 'Valid',\n",
      "                                                                          'originalState': 'Validating',\n",
      "                                                                          'submissionDate': {   'date': '2017-10-02T18:11:26.802+0000'}}],\n",
      "                                                        'submissionDate': {   'date': '2017-10-02T18:10:36.669Z'},\n",
      "                                                        'updateDate': {   'date': '2017-10-02T18:11:26.802Z'},\n",
      "                                                        'uuid': '7648ad7b-be2c-4515-9e2c-fa6d3c95baaf',\n",
      "                                                        'validationState': 'Valid'},\n",
      "                                            'description': 'Human primary '\n",
      "                                                           'islets (85-95% '\n",
      "                                                           'pure) were '\n",
      "                                                           'purchased from '\n",
      "                                                           'Prodo Laboratories '\n",
      "                                                           'Inc (Irvine, CA, '\n",
      "                                                           'USA) and cultured '\n",
      "                                                           'for 4 days in '\n",
      "                                                           'Prodo Islet Media '\n",
      "                                                           'Standard PIM(S) to '\n",
      "                                                           'recover after '\n",
      "                                                           'arrival. Islets '\n",
      "                                                           'were dissociated '\n",
      "                                                           'into single-cell '\n",
      "                                                           'suspension and '\n",
      "                                                           'viable individual '\n",
      "                                                           'cells were '\n",
      "                                                           'distributed by '\n",
      "                                                           'FACS into 384-well '\n",
      "                                                           'plates containing '\n",
      "                                                           'lysis buffer. '\n",
      "                                                           'Spike-in RNA from '\n",
      "                                                           'the External RNA '\n",
      "                                                           'Controls '\n",
      "                                                           'Consortium (ERCC) '\n",
      "                                                           '(Ambion, Life '\n",
      "                                                           'Technologies) was '\n",
      "                                                           'added to each well '\n",
      "                                                           'in 1:40 000 '\n",
      "                                                           'dilution. ',\n",
      "                                            'id': 'p1',\n",
      "                                            'type': 'sample collection '\n",
      "                                                    'protocol'},\n",
      "                                        {   'core': {   'accession': None,\n",
      "                                                        'events': [   {   'endState': 'Validating',\n",
      "                                                                          'originalState': 'Draft',\n",
      "                                                                          'submissionDate': {   'date': '2017-10-02T18:11:16.787+0000'}},\n",
      "                                                                      {   'endState': 'Valid',\n",
      "                                                                          'originalState': 'Validating',\n",
      "                                                                          'submissionDate': {   'date': '2017-10-02T18:11:29.945+0000'}}],\n",
      "                                                        'submissionDate': {   'date': '2017-10-02T18:10:36.678Z'},\n",
      "                                                        'updateDate': {   'date': '2017-10-02T18:11:29.945Z'},\n",
      "                                                        'uuid': 'a4796352-436e-4c25-ae40-ae416f7ffe14',\n",
      "                                                        'validationState': 'Valid'},\n",
      "                                            'description': 'Single-cell '\n",
      "                                                           'RNA-seq libraries '\n",
      "                                                           'were generated as '\n",
      "                                                           'described in '\n",
      "                                                           'Picelli et al., '\n",
      "                                                           '2014, Full-length '\n",
      "                                                           'RNA-seq from '\n",
      "                                                           'single cells using '\n",
      "                                                           'Smart-seq2, Nature '\n",
      "                                                           'Protocols. ',\n",
      "                                            'id': 'p2',\n",
      "                                            'type': 'nucleic acid library '\n",
      "                                                    'construction protocol'},\n",
      "                                        {   'core': {   'accession': None,\n",
      "                                                        'events': [   {   'endState': 'Validating',\n",
      "                                                                          'originalState': 'Draft',\n",
      "                                                                          'submissionDate': {   'date': '2017-10-02T18:11:16.795+0000'}},\n",
      "                                                                      {   'endState': 'Valid',\n",
      "                                                                          'originalState': 'Validating',\n",
      "                                                                          'submissionDate': {   'date': '2017-10-02T18:11:29.874+0000'}}],\n",
      "                                                        'submissionDate': {   'date': '2017-10-02T18:10:36.686Z'},\n",
      "                                                        'updateDate': {   'date': '2017-10-02T18:11:29.874Z'},\n",
      "                                                        'uuid': '8889df7f-5ceb-4ccf-951d-0ffad28543a6',\n",
      "                                                        'validationState': 'Valid'},\n",
      "                                            'description': 'Libraries were '\n",
      "                                                           'sequenced on an '\n",
      "                                                           'Illumina HiSeq '\n",
      "                                                           '2000, generating '\n",
      "                                                           '43 bp single-end '\n",
      "                                                           'reads.',\n",
      "                                            'id': 'p3',\n",
      "                                            'type': 'nucleic acid sequencing '\n",
      "                                                    'protocol'},\n",
      "                                        {   'core': {   'accession': None,\n",
      "                                                        'events': [   {   'endState': 'Validating',\n",
      "                                                                          'originalState': 'Draft',\n",
      "                                                                          'submissionDate': {   'date': '2017-10-02T18:11:16.788+0000'}},\n",
      "                                                                      {   'endState': 'Valid',\n",
      "                                                                          'originalState': 'Validating',\n",
      "                                                                          'submissionDate': {   'date': '2017-10-02T18:11:28.862+0000'}}],\n",
      "                                                        'submissionDate': {   'date': '2017-10-02T18:10:36.694Z'},\n",
      "                                                        'updateDate': {   'date': '2017-10-02T18:11:28.862Z'},\n",
      "                                                        'uuid': 'b58d1329-08b9-48a7-82dd-2df708504af8',\n",
      "                                                        'validationState': 'Valid'},\n",
      "                                            'description': 'Expression values '\n",
      "                                                           'for RefSeq genes '\n",
      "                                                           'and values for '\n",
      "                                                           'ERCC transcripts '\n",
      "                                                           'were quantified as '\n",
      "                                                           'RPKMs using '\n",
      "                                                           'rpkmforgenes '\n",
      "                                                           '(Ramsköld et al., '\n",
      "                                                           '2009). Format of '\n",
      "                                                           'the datafile '\n",
      "                                                           '‘pancreas_refseq_rpkms_counts_3514sc.txt’: '\n",
      "                                                           'The file contains '\n",
      "                                                           'both the '\n",
      "                                                           'normalized rpkm '\n",
      "                                                           'values and the raw '\n",
      "                                                           'read counts for '\n",
      "                                                           'each sample. '\n",
      "                                                           'Columns correspond '\n",
      "                                                           'to samples and '\n",
      "                                                           'rows to genes. The '\n",
      "                                                           'first line of the '\n",
      "                                                           'file (starting '\n",
      "                                                           'with: #samples) '\n",
      "                                                           'contains the '\n",
      "                                                           'sample IDs to be '\n",
      "                                                           'used as column '\n",
      "                                                           'labels for both '\n",
      "                                                           'the rpkm and '\n",
      "                                                           'counts. The '\n",
      "                                                           'columns of the '\n",
      "                                                           'rpkm and the '\n",
      "                                                           'counts have the '\n",
      "                                                           'same order with '\n",
      "                                                           'the sample IDs. '\n",
      "                                                           'Columns 1:3514 '\n",
      "                                                           'correspond to rpkm '\n",
      "                                                           'values, Columns '\n",
      "                                                           '3515:7028 '\n",
      "                                                           'correspond to read '\n",
      "                                                           'counts. Rows '\n",
      "                                                           '1:26179 correspond '\n",
      "                                                           'to data for RefSeq '\n",
      "                                                           'genes, Rows '\n",
      "                                                           '26180:26271 '\n",
      "                                                           'correspond to data '\n",
      "                                                           'for the 92 '\n",
      "                                                           'external RNA '\n",
      "                                                           'spike-in controls '\n",
      "                                                           '(ERCCs), Row 26272 '\n",
      "                                                           '(last) contains '\n",
      "                                                           'data for ‘eGFP’.',\n",
      "                                            'id': 'p4',\n",
      "                                            'type': 'normalization data '\n",
      "                                                    'transformation protocol'}],\n",
      "                       'supplementary_files': [   'ftp://ftp.ebi.ac.uk/pub/databases/microarray/data/experiment/MTAB/E-MTAB-5061/E-MTAB-5061.processed.1.zip'],\n",
      "                       'title': 'Pancreatic islets from male donor, age 23, '\n",
      "                                'BMI 21.5, no diabetes'}}\n",
      "\n",
      "#####################################################\n",
      "#                  PRINTING DATA FILES               #\n",
      "#####################################################\n",
      "{   'ERR1631082.fastq.gz': {   'content-type': 'hca-data-file',\n",
      "                               'crc32c': '4cc3addf',\n",
      "                               'indexed': False,\n",
      "                               'name': 'ERR1631082.fastq.gz',\n",
      "                               's3_etag': '54cee219a07b9a2017f7fef6754b9221',\n",
      "                               'sha1': '5e1c8321d7199d956caedf36c10ffb8afb6bb08e',\n",
      "                               'sha256': 'f54a17b0a70fa50a7a7d95cf1fa9b9d99449bb739b609eb812a7d12be796a752',\n",
      "                               'size': 26,\n",
      "                               'uuid': 'caf9ed7f-06e8-4013-9fcf-a367e3a0cad0',\n",
      "                               'version': '2017-10-02T183002.375323Z'}}\n"
     ]
    }
   ],
   "source": [
    "# Create DataExtractor instance pointing to HCA Staging\n",
    "extractor = DataExtractor(\"https://dss.staging.data.humancellatlas.org/v1\")\n",
    "# Use dummy payload and get the metadata_files and the data_files\n",
    "metadata_files, data_files = extractor.extract_bundle(payload, \"aws\")\n",
    "# Print each dictionary\n",
    "print(\"\\n#####################################################\")\n",
    "print(\"#                    PRINTING METADATA               #\")\n",
    "print(\"#####################################################\")\n",
    "pprint(metadata_files, indent=4)\n",
    "print(\"\\n#####################################################\")\n",
    "print(\"#                  PRINTING DATA FILES               #\")\n",
    "print(\"#####################################################\")\n",
    "pprint(data_files, indent=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we pass this on to the FileIndexer class to create a File Oriented index entry on ElasticSearch running on `localhost:9200`. But first, we get the index settings and get the configuration files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INDEXING DONE\n"
     ]
    }
   ],
   "source": [
    "# Define helper method to open files\n",
    "def open_and_return_json(file_path):\n",
    "    \"\"\"\n",
    "    Opens and returns the contents of the json file given in file_path\n",
    "    :param file_path: Path of a json file to be opened\n",
    "    :return: Returns an obj with the contents of the json file\n",
    "    \"\"\"\n",
    "    with open(file_path, 'r') as file_:\n",
    "        loaded_file = json.load(file_)\n",
    "    return loaded_file\n",
    "\n",
    "# Get the index's settings\n",
    "index_settings = open_and_return_json('chalicelib/settings.json')\n",
    "# Get the index overall config\n",
    "index_mapping_config = open_and_return_json('chalicelib/config.json')\n",
    "\n",
    "file_indexer = FileIndexer(metadata_files,\n",
    "                           data_files,\n",
    "                           es,\n",
    "                           \"testing_index\",\n",
    "                           \"doc\",\n",
    "                           index_settings=index_settings,\n",
    "                           index_mapping_config=index_mapping_config)\n",
    "\n",
    "file_indexer.index(bundle_uuid, bundle_version)\n",
    "print(\"INDEXING DONE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
